### PR TITLE
LASB-1973 - Implemented Retries for CDA Calls and Integration Tests

### PR DESCRIPTION
**Pull request by lkm0287**
## What

[Link to story](https://dsdmoj.atlassian.net/browse/LASB-1973)

Implemented retries for calls to CDA and relevant integration tests.

## Checklist

Before you ask people to review this PR:

- [ ] Tests should be passing: `./gradlew test`
- [ ] Github should not be reporting conflicts; you should have recently run `git rebase main`.
- [ ] Avoid mixing whitespace changes with code changes in the same commit. These make diffs harder to read and conflicts more likely.
- [ ] You should have looked at the diff against main and ensured that nothing unexpected is included in your changes.
- [ ] You should have checked that the commit messages say why the change was made.

### CODE REVIEW

File: maat-court-data-api/src/test/java/gov/uk/courtdata/courtdataadapter/client/CourtDataAdapterClientIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        CourtDataAdapterOAuth2ClientConfig courtDataAdapterOAuth2ClientConfig = new CourtDataAdapterOAuth2ClientConfig();
+        return courtDataAdapterOAuth2ClientConfig.webClient(baseUrl, authClientManager, MAX_RETRIES, MIN_BACKOFF_PERIOD, JITTER);
+    }
+
+    @Test
+    public void givenAValidLaaStatusObject_whenPostLaaStatusIsInvoked_andSomeErrorResponsesReceivedFromCDAAreUnderMaxRetries_thenTheRequestIsSentCorrectlyAndAnAcceptedResponseIsReceived() {
+        // given
+        courtDataAdapterClient = new CourtDataAdapterClient(webClient, gsonBuilder, queueMessageLogService, courtDataAdapterClientConfig);
+        when(courtDataAdapterClientConfig.getLaaStatusUrl()).thenReturn(LAA_STATUS_URL);
+        mockCourtDataAdapterApi.enqueue(new MockResponse().setResponseCode(HttpStatus.INTERNAL_SERVER_ERROR.value()));
```
Comment (mtac50) [inline maat-court-data-api/src/test/java/gov/uk/courtdata/courtdataadapter/client/CourtDataAdapterClientIntegrationTest.java:83]: Could we move the error sequencing into a common method? That way you can try the number of failed responses to the max retries value i.e. `setupMockApiResponses(MAX_RETRIES + 1);` for failures then  `setupMockApiResponses(MAX_RETRIES - 1);` for success?

```
private setupMockApiResponses(Integer attemptsBeforeSuccess) {
        for (int i = 0; i < attemptsBeforeSuccess; i++) {
            mockCourtDataAdapterApi.enqueue(new MockResponse()
                    .setResponseCode(HttpStatus.INTERNAL_SERVER_ERROR.value()));
        }

        mockCourtDataAdapterApi.enqueue(new MockResponse().setResponseCode(HttpStatus.ACCEPTED.value()));
    }
```

File: maat-court-data-api/src/main/java/gov/uk/courtdata/courtdataadapter/client/CourtDataAdapterClient.java [hunk 2]
```diff
@@ -2,6 +2,7 @@
 
 import com.google.gson.GsonBuilder;
 import gov.uk.courtdata.enums.MessageType;
+import gov.uk.courtdata.exception.ApiClientException;
```
Comment (mtac50) [inline maat-court-data-api/src/main/java/gov/uk/courtdata/courtdataadapter/client/CourtDataAdapterClient.java:5]: Can this be removed?

File: maat-court-data-api/build.gradle [hunk 3]
```diff
@@ -154,6 +154,9 @@ dependencies {
     testImplementation 'org.testcontainers:junit-jupiter:1.16.0'
     testImplementation 'org.awaitility:awaitility:4.2.0'
 
+    // for retry integration tests on Mac M1 hardware
```
Comment (mtac50) [inline maat-court-data-api/build.gradle:157]: As discussed I think we can remove this for now.

File: maat-court-data-api/src/test/java/gov/uk/courtdata/courtdataadapter/client/CourtDataAdapterClientIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        logger.addAppender(listAppender);
+    }
+
+    private WebClient buildWebClient() {
+        CourtDataAdapterOAuth2ClientConfig courtDataAdapterOAuth2ClientConfig = new CourtDataAdapterOAuth2ClientConfig();
+        return courtDataAdapterOAuth2ClientConfig.webClient(baseUrl, authClientManager, MAX_RETRIES, MIN_BACKOFF_PERIOD, JITTER);
+    }
+
+    @Test
+    public void givenAValidLaaStatusObject_whenPostLaaStatusIsInvoked_andSomeErrorResponsesReceivedFromCDAAreUnderMaxRetries_thenTheRequestIsSentCorrectlyAndAnAcceptedResponseIsReceived() {
```
Comment (mtac50) [inline maat-court-data-api/src/test/java/gov/uk/courtdata/courtdataadapter/client/CourtDataAdapterClientIntegrationTest.java:79]: Also SonarCloud would like the public modifiers removed from the tests!

File: maat-court-data-api/build.gradle (hunk 3 above)
Comment (lkm0287) [reply maat-court-data-api/build.gradle:157]: Done.

File: maat-court-data-api/src/main/java/gov/uk/courtdata/courtdataadapter/client/CourtDataAdapterClient.java (hunk 2 above)
Comment (lkm0287) [reply maat-court-data-api/src/main/java/gov/uk/courtdata/courtdataadapter/client/CourtDataAdapterClient.java:5]: Done.

File: maat-court-data-api/src/test/java/gov/uk/courtdata/courtdataadapter/client/CourtDataAdapterClientIntegrationTest.java (hunk 4 above)
Comment (lkm0287) [reply maat-court-data-api/src/test/java/gov/uk/courtdata/courtdataadapter/client/CourtDataAdapterClientIntegrationTest.java:79]: Done.

File: maat-court-data-api/src/test/java/gov/uk/courtdata/courtdataadapter/client/CourtDataAdapterClientIntegrationTest.java (hunk 1 above)
Comment (lkm0287) [reply maat-court-data-api/src/test/java/gov/uk/courtdata/courtdataadapter/client/CourtDataAdapterClientIntegrationTest.java]: Done.

**Review by mtac50 — APPROVED** [review]

**Review by alanrana7 — APPROVED** [review]

**Pull request merged**